internal/querycoordv2/balance/rowcount_based_balancer.go

```diff
@@ -0,0 +1,191 @@
+package balance
+
+import (
+	"sort"
+
+	"github.com/milvus-io/milvus/internal/querycoordv2/meta"
+	"github.com/milvus-io/milvus/internal/querycoordv2/session"
+	"github.com/milvus-io/milvus/internal/querycoordv2/task"
+)
+
+type RowCountBasedBalancer struct {
+	RoundRobinBalancer
+	nodeManager *session.NodeManager
+	dist        *meta.DistributionManager
+	meta        *meta.Meta
+}
+
+func (b *RowCountBasedBalancer) AssignSegment(segments []*meta.Segment, nodes []int64) []SegmentAssignPlan {
+	if len(nodes) == 0 {
+		return nil
+	}
+	nodeItems := b.convertToNodeItems(nodes)
+	queue := newPriorityQueue()
+	for _, item := range nodeItems {
+		queue.push(item)
+	}
+
+	sort.Slice(segments, func(i, j int) bool {
+		return segments[i].GetNumOfRows() > segments[j].GetNumOfRows()
+	})
+
+	plans := make([]SegmentAssignPlan, 0, len(segments))
+	for _, s := range segments {
+		// pick the node with the least row count and allocate to it.
+		ni := queue.pop().(*nodeItem)
+		plan := SegmentAssignPlan{
+			From:    -1,
+			To:      ni.nodeID,
+			Segment: s,
+		}
+		plans = append(plans, plan)
+		// change node's priority and push back
+		p := ni.getPriority()
+		ni.setPriority(p + int(s.GetNumOfRows()))
+		queue.push(ni)
+	}
+	return plans
+}
+
+func (b *RowCountBasedBalancer) convertToNodeItems(nodeIDs []int64) []*nodeItem {
+	ret := make([]*nodeItem, 0, len(nodeIDs))
+	for _, node := range nodeIDs {
+		segments := b.dist.SegmentDistManager.GetByNode(node)
+		rowcnt := 0
+		for _, s := range segments {
+			rowcnt += int(s.GetNumOfRows())
+		}
+		// more row count, less priority
+		nodeItem := newNodeItem(rowcnt, node)
+		ret = append(ret, &nodeItem)
+	}
+	return ret
+}
+
+func (b *RowCountBasedBalancer) Balance() ([]SegmentAssignPlan, []ChannelAssignPlan) {
+	ids := b.meta.CollectionManager.GetAll()
+
+	segmentPlans, channelPlans := make([]SegmentAssignPlan, 0), make([]ChannelAssignPlan, 0)
```

Comment:
is there reasonable to balance segment and channel together?

Reply:
> is there reasonable to balance segment and channel together?

I think yes. It is a interface to do a overall balance

Reply:
might channel and segment balance tasks have different priorities?